diff --git a/src/web/frontend.rs b/src/web/frontend.rs
index c31b5ac..102b0de 100644
--- a/src/web/frontend.rs
+++ b/src/web/frontend.rs
@@ -4,12 +4,14 @@ use axum::{
 };
 use reqwest::header;
 use rust_embed::RustEmbed;
+use std::{env, sync::OnceLock};
 
 const INDEX_HTML: &str = "index.html";
 
 #[derive(RustEmbed)]
 #[folder = "$CARGO_MANIFEST_DIR/web/dist"]
 struct Assets;
+static INDEX_HTML_CACHED: OnceLock<String> = OnceLock::new();
 
 pub async fn static_asset(uri: Uri) -> Response {
     let path = uri.path().trim_start_matches('/');
@@ -33,11 +35,32 @@ pub async fn static_asset(uri: Uri) -> Response {
     }
 }
 
+fn analytics_script() -> Option<String> {
+    let url = env::var("ANALYTICS_URL").ok()?;
+    let uuid = env::var("ANALYTICS_UUID").ok()?;
+
+    Some(format!(
+        r#"<script defer src="{url}" data-website-id="{uuid}"></script>"#
+    ))
+}
+
 async fn index_html() -> Response {
-    match Assets::get(INDEX_HTML) {
-        Some(content) => ([(header::CONTENT_TYPE, "text/html")], content.data).into_response(),
-        None => not_found().await,
-    }
+    let cached_html = INDEX_HTML_CACHED.get_or_init(|| {
+        let content = Assets::get(INDEX_HTML).expect("INDEX_HTML not found");
+        let mut html = String::from_utf8(content.data.to_vec()).expect("Invalid UTF-8");
+        
+        if let Some(script) = analytics_script() {
+            if html.contains("</head>") {
+                html = html.replace("</head>", &format!("{script}\n</head>"));
+            }
+        }
+        html
+    });
+
+    (
+        [(header::CONTENT_TYPE, "text/html; charset=utf-8")],
+        cached_html.clone(),
+    ).into_response()
 }
 
 async fn not_found() -> Response {
diff --git a/src/web/mod.rs b/src/web/mod.rs
index 207ee6a..c454878 100644
--- a/src/web/mod.rs
+++ b/src/web/mod.rs
@@ -143,7 +143,7 @@ pub async fn run(app: App, mut shutdown_rx: ShutdownRx, bot_tx: Sender<BotMessag
                 op.description("Get user logs in a channel from the given month")
             }),
         )
-        .api_route("/optout", post(handlers::optout))
+        //.api_route("/optout", post(handlers::optout))
         .api_route("/capabilities", get(capabilities))
         .route("/docs", Scalar::new("/openapi.json").axum_route())
         .route("/openapi.json", get(serve_openapi))
